services/src/main/java/org/keycloak/protocol/oidc/grants/management/endpoints/GrantEndpoint.java

```diff
@@ -0,0 +1,321 @@
+ *  * See the License for the specific language governing permissions and
+ *  * limitations under the License.
+ *
+ */
+
+package org.keycloak.protocol.oidc.grants.management.endpoints;
+
+import org.apache.commons.lang.StringUtils;
+import org.jboss.logging.Logger;
+import org.jboss.resteasy.annotations.cache.NoCache;
+import org.jboss.resteasy.spi.HttpRequest;
+import org.jboss.resteasy.spi.HttpResponse;
+import org.keycloak.OAuthErrorException;
+import org.keycloak.TokenVerifier;
+import org.keycloak.common.ClientConnection;
+import org.keycloak.common.VerificationException;
+import org.keycloak.common.constants.ServiceAccountConstants;
+import org.keycloak.crypto.SignatureProvider;
+import org.keycloak.crypto.SignatureVerifierContext;
+import org.keycloak.events.Details;
+import org.keycloak.events.Errors;
+import org.keycloak.events.EventBuilder;
+import org.keycloak.events.EventType;
+import org.keycloak.models.*;
+import org.keycloak.models.utils.KeycloakModelUtils;
+import org.keycloak.protocol.oidc.OIDCAdvancedConfigWrapper;
+import org.keycloak.protocol.oidc.OIDCLoginProtocol;
+import org.keycloak.protocol.oidc.TokenManager;
+import org.keycloak.representations.AccessToken;
+import org.keycloak.services.CorsErrorResponseException;
+import org.keycloak.services.Urls;
+import org.keycloak.services.managers.AppAuthManager;
+import org.keycloak.services.managers.AuthenticationManager;
+import org.keycloak.services.managers.UserSessionCrossDCManager;
+import org.keycloak.services.resource.RealmResourceProvider;
+import org.keycloak.services.resources.Cors;
+import org.keycloak.services.util.MtlsHoKTokenUtil;
+import org.keycloak.sessions.AuthenticationSessionModel;
+import org.keycloak.sessions.RootAuthenticationSessionModel;
+
+import javax.ws.rs.*;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.Collections;
+
+/**
+ * Grant management for OAuth 2.0 endpoint
+ */
+public class GrantEndpoint implements RealmResourceProvider {
+
+    public static final String GRANT_MANAGEMENT_QUERY = "grant_management_query";
+    public static final String GRANT_MANAGEMENT_REVOKE = "grant_management_revoke";
+
+    protected static final Logger logger = Logger.getLogger(GrantEndpoint.class);
+
+    @Context
+    private HttpRequest request;
+
+    @Context
+    private HttpResponse response;
+
+    @Context
+    private KeycloakSession session;
+
+    @Context
+    private ClientConnection clientConnection;
+
+    private final org.keycloak.protocol.oidc.TokenManager tokenManager;
+    private final AppAuthManager appAuthManager;
+    private final RealmModel realm;
+    private ClientModel clientModel;
+    private UserModel userModel;
+    private EventBuilder event;
+    private Cors cors;
+
+
+
+
+    public GrantEndpoint(RealmModel realm, EventBuilder event, org.keycloak.protocol.oidc.TokenManager tokenManager) {
```

Comment:
IMO, it might be better to authenticate the client who send grant management request (query, revoke) with an access token.
Moreover, it might better to confirm that the authenticated client is the one to whom keycloak issued the access token.

The current endpoint allows only simple bearer token usage so that anyone who steals other    legitimate client's access token can revoke their grants.

Reply:
agree